File: HRMS/HRMS.Web/Modules/ServerTypes/Operations/LeaveForm.ts
import { DateEditor, DecimalEditor, EnumEditor, ImageUploadEditor, initFormType, LookupEditor, PrefixedContext, StringEditor, TextAreaEditor } from "@serenity-is/corelib";
import { HalfDaySession } from "./HalfDaySession";
import { HrApprovalStatus } from "./HrApprovalStatus";
import { LeaveFinalStatus } from "./LeaveFinalStatus";
import { LeaveStatus } from "./LeaveStatus";

export interface LeaveForm {
    LeaveApplicationNo: StringEditor;
    ApplicationDate: DateEditor;
    EmployeeId: LookupEditor;
    LeaveTypeId: LookupEditor;
    StartDate: DateEditor;
    EndDate: DateEditor;
    HalfDaySession: EnumEditor;
    TotalDays: DecimalEditor;
    PaidDays: DecimalEditor;
    UnpaidDays: DecimalEditor;
    Reason: TextAreaEditor;
    Attachment: ImageUploadEditor;
    ReportingManagerId: LookupEditor;
    Status: EnumEditor;
    HrApprovalStatus: EnumEditor;
    FinalStatus: EnumEditor;
    ManagerRemarks: TextAreaEditor;
    HrRemarks: TextAreaEditor;
    SubstituteEmployeeId: LookupEditor;
    ContactDuringLeave: StringEditor;
}

export class LeaveForm extends PrefixedContext {
    static readonly formKey = 'Operations.Leave';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!LeaveForm.init) {
            LeaveForm.init = true;

            var w0 = StringEditor;
            var w1 = DateEditor;
            var w2 = LookupEditor;
            var w3 = EnumEditor;
            var w4 = DecimalEditor;
            var w5 = TextAreaEditor;
            var w6 = ImageUploadEditor;

            initFormType(LeaveForm, [
                'LeaveApplicationNo', w0,
                'ApplicationDate', w1,
                'EmployeeId', w2,
                'LeaveTypeId', w2,
                'StartDate', w1,
                'EndDate', w1,
                'HalfDaySession', w3,
                'TotalDays', w4,
                'PaidDays', w4,
                'UnpaidDays', w4,
                'Reason', w5,
                'Attachment', w6,
                'ReportingManagerId', w2,
                'Status', w3,
                'HrApprovalStatus', w3,
                'FinalStatus', w3,
                'ManagerRemarks', w5,
                'HrRemarks', w5,
                'SubstituteEmployeeId', w2,
                'ContactDuringLeave', w0
            ]);
        }
    }
}

[HalfDaySession, LeaveStatus, HrApprovalStatus, LeaveFinalStatus]; // referenced types

File: HRMS/HRMS.Web/Modules/Operations/Leave/LeaveForm.cs
using Serenity.ComponentModel;

namespace HRMS.Operations.Forms;

[FormScript("Operations.Leave")]
[BasedOnRow(typeof(LeaveRow), CheckNames = true)]
public class LeaveForm
{
    [ReadOnly(true)]
    public string LeaveApplicationNo { get; set; }
    [ReadOnly(true)]
    public DateTime ApplicationDate { get; set; }
    public int EmployeeId { get; set; }
    public int LeaveTypeId { get; set; }
    [HalfWidth]
    public DateTime StartDate { get; set; }
    [HalfWidth]
    public DateTime EndDate { get; set; }
    public HalfDaySession HalfDaySession { get; set; }
    [ReadOnly(true)]
    public double TotalDays { get; set; }
    [ReadOnly(true)]
    public decimal PaidDays { get; set; }
    [ReadOnly(true)]
    public decimal UnpaidDays { get; set; }
    [TextAreaEditor(Rows = 3)]
    public string Reason { get; set; }
    public string Attachment { get; set; }
    public int ReportingManagerId { get; set; }
    [ReadOnly(true)]
    public LeaveStatus Status { get; set; }
    [ReadOnly(true)]
    public HrApprovalStatus HrApprovalStatus { get; set; }
    [ReadOnly(true)]
    public LeaveFinalStatus FinalStatus { get; set; }
    [TextAreaEditor(Rows = 2)]
    public string ManagerRemarks { get; set; }
    [TextAreaEditor(Rows = 2)]
    public string HrRemarks { get; set; }
    public int SubstituteEmployeeId { get; set; }
    public string ContactDuringLeave { get; set; }
}
